File: temporal/worker-go/activities/event_activities.go
package activities

import (
	"context"
	"fmt"
	"time"

	"go.temporal.io/sdk/activity"
)

// ValidateEvent checks that the event is well-formed.
func ValidateEvent(ctx context.Context, eventType string, eventID string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("ValidateEvent", "type", eventType, "id", eventID)

	// Simulate validation work
	time.Sleep(500 * time.Millisecond)

	if eventType == "" || eventID == "" {
		return "", fmt.Errorf("invalid event: type and id are required")
	}

	return fmt.Sprintf("validated:%s:%s", eventType, eventID), nil
}

// ProcessEvent performs the main processing logic for the event.
func ProcessEvent(ctx context.Context, eventType string, eventID string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("ProcessEvent", "type", eventType, "id", eventID)

	// Simulate processing work
	time.Sleep(1 * time.Second)

	return fmt.Sprintf("processed:%s:%s", eventType, eventID), nil
}

// NotifyEventProcessed sends a notification that event processing is complete.
func NotifyEventProcessed(ctx context.Context, eventType string, eventID string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("NotifyEventProcessed", "type", eventType, "id", eventID)

	// Simulate sending a notification (email, webhook, etc.)
	time.Sleep(300 * time.Millisecond)

	return fmt.Sprintf("notified:%s:%s", eventType, eventID), nil
}

File: temporal/worker-ts/src/activities/eventActivities.ts
import { log } from '@temporalio/activity';

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function validateEvent(eventType: string, eventId: string): Promise<string> {
  log.info('validateEvent', { type: eventType, id: eventId });
  await delay(500);

  if (!eventType || !eventId) {
    throw new Error('Invalid event: type and id are required');
  }

  return `validated:${eventType}:${eventId}`;
}

/**
 * enrichEvent — TS-specific activity that runs in parallel with validateEvent.
 * Simulates looking up metadata or enriching the event payload.
 */
export async function enrichEvent(eventType: string, eventId: string): Promise<string> {
  log.info('enrichEvent', { type: eventType, id: eventId });
  await delay(700); // Slightly longer than validate to show parallel benefit

  return `enriched:${eventType}:${eventId}`;
}

export async function processEvent(eventType: string, eventId: string): Promise<string> {
  log.info('processEvent', { type: eventType, id: eventId });
  await delay(1000);

  return `processed:${eventType}:${eventId}`;
}

export async function notifyEventProcessed(eventType: string, eventId: string): Promise<string> {
  log.info('notifyEventProcessed', { type: eventType, id: eventId });
  await delay(300);

  return `notified:${eventType}:${eventId}`;
}
